server/src/main/java/io/opencmw/server/MajordomoWorker.java

```diff
@@ -141,16 +145,28 @@ protected MajordomoWorker(final ZContext ctx, final String brokerAddress, final
                 serialiser.getMatchedIoSerialiser().setBuffer(defaultBuffer);
                 serialiser.serialiseObject(output);
                 defaultBuffer.flip();
-                c.rep.data = Arrays.copyOf(defaultBuffer.elements(), defaultBuffer.limit() + 4); //TODO: investigate why we need 4 bytes of buffer at the end
+                c.rep.data = Arrays.copyOf(defaultBuffer.elements(), defaultBuffer.limit() + 4);
+                break;
+            case CMWLIGHT:
+                serialiser.setMatchedIoSerialiser(CmwLightSerialiser.class);
+                serialiser.getMatchedIoSerialiser().setBuffer(defaultBuffer);
+                serialiser.serialiseObject(output);
+                defaultBuffer.flip();
+                c.rep.data = Arrays.copyOf(defaultBuffer.elements(), defaultBuffer.limit());
                 break;
             case BINARY:
-            case UNKNOWN:
-            default:
                 serialiser.setMatchedIoSerialiser(BinarySerialiser.class);
                 serialiser.getMatchedIoSerialiser().setBuffer(defaultBuffer);
                 serialiser.serialiseObject(output);
                 defaultBuffer.flip();
-                c.rep.data = Arrays.copyOf(defaultBuffer.elements(), defaultBuffer.limit() + 4); //TODO: investigate why we need 4 bytes of buffer at the end
+                c.rep.data = Arrays.copyOf(defaultBuffer.elements(), defaultBuffer.limit());
+                break;
+            case UNKNOWN:
+            default:
+                if (output instanceof BinaryData) {
+                    c.rep.data = ((BinaryData) output).data;
+                }
+                // return c.rep.data as defined in the user handler
```

Comment:
I'm not sure if returning empty data by default is the best choice. This code path just silently discards the object the user has supplied.
On the other hand if the user has explicitly set `c.rep.data` we should probably not overwrite it? Maybe it would be better to skip serialisation in that case and default to one of the serializers otherwise?
Also for BinaryData serialisation should be skipped, but the contentType should be set to the contentType of the BinaryData?

Reply:
That's the purpose of this UNKNOWN/default branch: to allow the user to return custom binary blobs with arbitrary content. The `if (output instanceof BinaryData)` is just syntactic sugar whether the user wants to set via `c.rep.data` or via the explicitly declared `BinaryData` return object. This is also the answer to the question above why there is no default return object in case there is no content type specified.... the default return is/has to be defined by the user implementing the property/service. ... to be further discussed...